multi-pool: add support for excluding IP pools from masquerade

hassenius opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

This adds a new flag only-masquerade-default-pool which disables masquerading for pods which have IPs allocated non-default pools.

A per-pool annotation (ipam.cilium.io/skip-masquerade) is also added to configure this behaviour in a more granular way.

This can be particularly useful when using the newly supported combination of multi-pool ipam and tunneling, where you may have the default pool be tunnelled, and additional pools handled natively (for example in conjunction with bgp control plane)

This feature is supported with a new stateDB table for CiliumPodIPPools. At IP allocation time, the chosen pool is looked up on stateDB to look for the skip-masquerade annotation.

Fixes: #40131
<!-----
```release-note

```
----->

bimmlerd commented on pkg/option/config.go:
```diff
@@ -3188,6 +3206,28 @@ func (c *DaemonConfig) checkIPv4NativeRoutingCIDR() error {
 		IPAM, c.IPAMMode())
 }
 
+func (c *DaemonConfig) checkMasqueradeSrcExclusionCIDR() error {
```
let's avoid adding more code to `option/config`. It may be possible to add the new config option to the iptables manager's `Config` struct (see https://docs.cilium.io/en/latest/contributing/development/hive/#api-config for docs on how this works) and have the validation also happen there, but you also need to change how you pass it to the DP so it depends a bit on that where the flag ought to live.

bimmlerd commented on pkg/datapath/linux/config/config.go:
```diff
@@ -582,6 +582,14 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 				}
+
+				// Add source exclusion CIDR if configured
+				if option.Config.MasqueradeSrcExclusionCIDR != nil && option.Config.MasqueradeSrcExclusionCIDR.IP.To4() != nil {
```
this way of doing configuration is deprecated; see above

hassenius commented:
Thanks for the review so far @bimmlerd 

I've made an attempt at moving the configuration into the iptables cell as requested, and I've updated the iptables implementation such that it adds a rule at the top of the `CILIUM_POST_nat` chain. This also opens the possibility of support an array (i.e. comma separated list of CIDR) as input, which would be useful for this feature, but not a pattern I remember having seen in cilium yet. Would appreciate thoughts on this.

For the bpf implementation I'm unsure how to pass this. Would really appreciate some suggestions.

alimehrabikoshki commented:
Hi @bimmlerd 

I added support for multiple CIDRs to the flag (comma-separated) and changed the BPF implementation to use LPM trie maps.

I'd appreciate a review on the way the config is done now, and where the map update logic is put in loader.go. 

The ip-masq-agent BPF map lookup was the same as ours so I made a common function for that as well.

I've also found that by default the iptables implementation doesn't masquerade the additional IP pools, but leaving in our iptables implementation anyway to be more explicit about it. 

Thanks

hahasheminejad commented:
> > Are there any known workarounds to achieve this behavior in the meantime? Also, is there an estimated timeline for when this PR might be merged or released?
> 
> The workaround we will use until this is merged is to revert back to iptables masq - so no BPF-based masquerade. Since this is a new feature I imagine it will go into 1.19.

Thank you.

We have come up with this nonMasqueradeCIDR list to exclude our internal subnet (10.0.0.0/8). It's been working fine so far 😊

```
nonMasqueradeCIDRs:
      - "0.0.0.0/5"
      - "8.0.0.0/7"
      - "11.0.0.0/8"
      - "12.0.0.0/6"
      - "16.0.0.0/4"
      - "32.0.0.0/3"
      - "64.0.0.0/2"
      - "128.0.0.0/1"
```

joestringer commented on pkg/datapath/iptables/cell.go:
```diff
@@ -82,6 +90,8 @@ func (def Config) Flags(flags *pflag.FlagSet) {
 	flags.Bool("iptables-random-fully", def.IPTablesRandomFully, "Set iptables flag random-fully on masquerading rules")
 	flags.Bool("prepend-iptables-chains", def.PrependIptablesChains, "Prepend custom iptables chains instead of appending")
 	flags.Bool("enable-xt-socket-fallback", def.EnableXTSocketFallback, "Enable fallback for missing xt_socket module")
+	flags.String("ipv4-masquerade-src-exclusion-cidrs", def.IPv4MasqueradeSrcExclusionCIDRs, "IPv4 CIDR for source IPs that should be excluded from masquerading")
+	flags.String("ipv6-masquerade-src-exclusion-cidrs", def.IPv6MasqueradeSrcExclusionCIDRs, "IPv6 CIDR for source IPs that should be excluded from masquerading")
```
Just noting that we have a lot of flags for masquerade and I think many people find them quite confusing. I had filed https://github.com/cilium/cilium/discussions/36335 to try to better understand the solution space and where we think the project should go in terms of this configuration, but at least right now I don't have specific solutions in mind. 

The one other thing that does come to mind is that the "native routing CIDR" is intended to describe which IP range that Cilium-managed Endpoints can directly communicate with, while avoiding SNAT. What's not clear to me is whether there is a meaningful semantic difference with these newly proposed flags or if they're trying to configure the same underlying functionality. I realize there's only one native routing CIDR at the moment but if that's the problem then maybe we could look at expanding that to accept a slice.

alimehrabikoshki replied:
We would use this in conjunction with BGP control plane to ensure these IPs are globally routable outside of the cluster network, so that no masquerading is required. Keeping the source IP even outside the cluster is the key attribute that enables our use cases.

I also theorized in https://github.com/cilium/cilium/issues/40131 that an annotation on the IPPool CR could configure this behavior, but it was a bit harder to implement. In the context of your worries about the masquerade configuration complexity, the annotation approach is certainly more opaque than being explicit about the CIDRs or your proposed long-term approach.

alimehrabikoshki replied:
> One of the consequences with adding yet more flags is that it can introduce an overlap with the other flags

That’s a valid concern. I’d certainly be lying if I said I’ve thought through all the different interactions with the other flags (even getting to the point of knowing our desired behavior is unsupported took a long time of fiddling with the different flags).

joestringer replied:
Yeah I think that's a natural progression of these settings, we're learning as we go along. It would certainly be nice to have some thought / writeup on how we think they should interact to have some sane behavior. Alternatively if for your use case these other settings are not necessary, we could add some checks on startup to prevent users from configuring multiple of these flags. This is not quite as nice as having an opinionated idea about how they should work, but particularly if the use cases are discrete it might help to avoid others getting quite as confused about how the flags should work.

alimehrabikoshki replied:
Yeah that makes sense. I can spend some time thinking about it and update this MR with my findings. Then we can decide if we want to disable the interaction with the other flags as you said.

alimehrabikoshki replied:
Had a look at the different flags (let me know if I missed any):

By default, traffic leaving the cluster is masqueraded (enable-ipv4/6-masquerade).

We can opt-out of masquerading based on:
1. Destination IP
ipv4|6-native-routing-cidr or enable-ip-masq-agent - these set destination CIDRs that should not be masqueraded
2. Source IP
New snat-exclusion-flags - if source IP is in the given CIDR, it will not masquerade.

We could match either or both and things would work as expected, as they both disable masquerading, so no conflict there.

Exact masquerade behaviour is further customized by:
1. Interfaces
egress-masquerade-interfaces - this limits masquerading to given interfaces (takes precedence over `devices`). This is compatible with snat-exclusion since they both try to limit when masquerading happens.
2. Routes
enable-masquerade-to-route-source=true uses the routing layer to select source IP to masquerade, but does not make decisions on whether to masq or not. This could be used in conjunction with snat-exclusion since snat-exclusion comes first and enable-masquerade-to-route-source only applies to masqueraded traffic.
3. Datapath
to use BPF or not, via enable-bpf-masquerade. We support both in this MR.


So the way I see it, these flags can be dropped in and configured with any other flag as needed.

alimehrabikoshki replied:
@joestringer does this look good to you?

joestringer replied:
There's one other flag, but I see this one as falling under the "Destination IP" category as a modifier for the ip-masq-agent behavior for better behavioral compatibility to upstream k8s ip-masq-agent: https://github.com/cilium/cilium/pull/37568 .

hassenius commented:
> Marking as draft while the above feedback is addressed.

Think everything has been addressed now thank you

joestringer commented:
Next steps: @cilium/cli and @cilium/proxy codeowners should review, and also this PR needs to be rebased against the tip of `main` to resolve conflicts, then we can run the testsuite.

aanm commented:
/test

bimmlerd commented:
I think my CODEOWNERS are no longer touched; removing my request for review.

bimmlerd commented on pkg/datapath/loader/loader.go:
```diff
@@ -320,6 +321,9 @@ func removeObsoleteNetdevPrograms(logger *slog.Logger, devices []string) error {
 // cilium_net and external (native) devices.
 func reloadHostEndpoint(logger *slog.Logger, ep datapath.Endpoint, lnc *datapath.LocalNodeConfiguration, spec *ebpf.CollectionSpec) error {
 	// Replace programs on cilium_host.
+	if err := reloadNatExclusionMaps(logger, lnc); err != nil {
```
nit: Go has the initialisms rule; so throughout it should be `NAT` or `nat` but never `Nat`: https://go.dev/wiki/CodeReviewComments#initialisms

alimehrabikoshki replied:
Thanks, fixed

bimmlerd commented on pkg/datapath/loader/loader.go:
```diff
@@ -335,6 +339,59 @@ func reloadHostEndpoint(logger *slog.Logger, ep datapath.Endpoint, lnc *datapath
+		return fmt.Errorf("failed to initialize NAT exclusion maps: %w", err)
+	}
+
+	v4ExclusionCidrs := lnc.GetIPv4MasqueradeSrcExclusionCIDRs()
```
nit: similar to nat above, `Cidrs` -> `CIDRs`

bimmlerd commented on pkg/maps/nat/exclusion.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+const (
+	NatExclusionMapNameIPv4 = "cilium_nat_exclusion_v4"
```
nit: this could just be `ExclusionMapNameIPv4`: https://go.dev/wiki/CodeReviewComments#package-names - usages in other packages otherwise "stutter"; `nat.Nat...`

Applies throughout this pkg

alimehrabikoshki replied:
Good shout, fixed

bimmlerd commented on pkg/maps/nat/exclusion.go:
```diff
@@ -0,0 +1,54 @@
+	NatExclusionMapIPv4 *cidrmap.CIDRMap
+	NatExclusionMapIPv6 *cidrmap.CIDRMap
+
+	initOnce sync.Once
```
package level state make it hard to test - can't run multiple tests which use this package (even transitively). Can we encapsulate this state in a struct?

alimehrabikoshki replied:
Hi @bimmlerd - is there anything I should do on this PR? I see you requested changes and a review from me, is review by the other CODEOWNERs blocked until I press the "Add your review" button? Your changes are all addressed I believe.

bimmlerd reviewed (changes requested):
Removing my request for review just reverted it back to a comment; so I just reviewed properly instead. I'm not a CODEOWNER of any of this though, so take my input with a grain of salt.

mhofstetter reviewed (approved):
approved. similar to @bimmlerd - looks like the latest changes removed the need for proxy & cli review (even though still in the reviewers list - but no longer owning any files touched by this PR).

joestringer commented:
/test

dylandreimerink commented:
@hassenius seems there are some merge conflicts, could you rebase on main and resolve them?

joestringer commented:
@alimehrabikoshki thanks for the report.. @aanm seems like this is related to https://github.com/cilium/cilium/pull/41681. Thoughts on how we should make this work? Problem is that new contributions require approval from a committer to avoid abuse, but that means the CI image build will always fail first time. We can approve but then we have to manually restart the Image CI build workflow as well. Is that the intended workflow or can we come up with something smoother?

joestringer commented:
/test

alimehrabikoshki commented:
I've rebased to resolve conflicts. @gentoo-root and @derailed, do you have any further concerns on this PR?

alimehrabikoshki commented:
> Latest rebase seems to hit some linting problem: https://github.com/cilium/cilium/actions/runs/19507355062/job/55840085730?pr=40132
> 
> ```
>  pkg/datapath/maps/cell.go:1: : # github.com/cilium/cilium/pkg/datapath/maps [github.com/cilium/cilium/pkg/datapath/maps.test]
>   Error: pkg/datapath/maps/map_test.go:239:13: undefined: NewMapSweeper
>   Error: pkg/datapath/maps/map_test.go:257:13: undefined: NewMapSweeper (typecheck)
> ```

Sorry, I should have run the tests when rebasing

>The behavior differs from the iptables masquerade mode, which is a bit odd but we can follow up on that separately.

I'm looking into this some more and I'll create an issue for it

tommyp1ckles commented on pkg/datapath/iptables/cell.go:
```diff
@@ -74,14 +74,24 @@ type Config struct {
 	EnableXTSocketFallback bool
+
+	// IPv4MasqueradeSrcExclusionCIDRs is a comma-separated list of IPv4 source CIDRs that should be excluded
+	// from masquerade.
```
nit: this specifically refers to "bpf-masquerade", correct?

julianwiedmann commented:
ack with what @brb said - if the use-case is primarily about Multi-Pool integration (cc @cilium/sig-ipam), then we should aim for a better user-experience than yet another knob & config option for the SNAT feature.

Some other high-level thoughts:
* I'm not convinced that we need to continue adding additional features in the iptables masq mode. In my opinion the development & maintenance effort is better spent on improving the BPF masq code.
* we clearly need to cover this enhancement in `Documentation/`. I believe that would also help to iron out the intended use-case, and expected behavior.
* The BPF masq code is already looking up the local endpoint. What if that map entry would instead tell us whether the the endpoint should be excluded from Masquerading? Similar as the heuristic we currently apply for host endpoints.

kglushen commented:
Hello, everyone!

Apologies for interrupting the discussion, but haven't you consider behavior similar to Calico pools?
There is option natOutgoing within pool configuration. When enabled, packets sent from Calico networked containers in this pool to destinations outside of any Calico IP pools will be masqueraded.

So it looks more useful and obvious than separate subnet lists.

kglushen commented:
> we have plans to utilise multi-pool IPAM with BGP control plane

Looks like we have the same aims, especially if you have IPv6 setup. :)
Pool-based configuration would be handier and more straightforward.

alimehrabikoshki commented:
> So if you have ideas how this functionality should be **properly** integrated, please explore them. And then let's do this once, and not merge something now when we already understand that it needs to be reworked later.

I’m kicking myself for giving off such an impression. I admit of course that these changes are driven by my own use cases, but I certainly wasn’t trying to suggest an implementation that would later be removed or completely reworked. 

I see going from a global pool masquerade config to a per-pool config as an evolution of the same feature. All the work done to enable the former would be reused by latter in a more specific way, and I imagined both ways to configure them are ultimately kept. I am open to your thoughts - you might prefer all of it implemented in one MR, or you have preference for just one of the ways. But no imposition here, just being upfront about my own goals.

alimehrabikoshki committed:
multi-pool: add support for excluding IP pools from masquerade

This adds a new flag only-masquerade-default-pool which disables masquerading for pods which have IPs allocated non-default pools.

A per-pool annotation (ipam.cilium.io/skip-masquerade) is also added to configure this behaviour in a more granular way.

This can be particularly useful when using the newly supported combination of multi-pool ipam and tunneling, where you may have the default pool be tunnelled, and additional pools handled natively (for example in conjunction with bgp control plane)

This feature is supported with a new stateDB table for CiliumPodIPPools. At IP allocation time, the chosen pool is looked up on stateDB to look for the skip-masquerade annotation.

Signed-off-by: alimehrabikoshki <alimehrabi97@gmail.com>

alimehrabikoshki commented:
@julianwiedmann I've re-done the PR to implement both configuration knobs that I mentioned, `--only-masquerade-default-pool` as a global configuration to skip masquerade for all non-default pools, and `ipam.cilium.io/skip-masquerade` as a per-pool annotation. I've not run it on a cluster yet to test (waiting for the Push dev charts CI to run) but I'd love to hear some feedback on the general direction.

fristonio reviewed (commented):
cilium/endpoint: Looks good 🚀 
One question to clarify my understanding of the behavior.

julianwiedmann commented on bpf/tests/no_snat_endpoint_masquerade_skip_test.c:
```diff
@@ -0,0 +1,111 @@
+#define ENABLE_BPF_MASQUERADE 1
+#define ENABLE_MASQUERADE_IPV4 1
+#define ENABLE_MASQUERADE_IPV6 1
+#define IS_BPF_HOST 1
```
I think the covered scenarios would also nicely fit as additional tests inside `bpf/tests/host_bpf_masq.c`, validating that the src IP is not re-written for such endpoints.

julianwiedmann commented on bpf/lib/eps.h:
```diff
@@ -31,6 +31,8 @@ struct endpoint_key {
 #define ENDPOINT_F_HOST			1 /* Special endpoint representing local host */
 #define ENDPOINT_F_ATHOSTNS		2 /* Endpoint located at the host networking namespace */
+#define ENDPOINT_F_NO_SNAT_V4	4 /* Endpoint should not be masqueraded for IPv4 */
+#define ENDPOINT_F_NO_SNAT_V6	8 /* Endpoint should not be masqueraded for IPv6 */
```
Unclear whether we need two flags for this on the datapath level.  Are we sharing endpoint map entries for an endpoint, so that the lookup for an endpoint's IPv4 and IPv6 address would return the **same** struct? That would feel rather odd.

julianwiedmann commented on pkg/ipam/types.go:
```diff
@@ -53,6 +56,9 @@ type AllocationResult struct {
 	// This is only useful in ENI mode.
 	InterfaceNumber string
+
+	// SkipMasquerade indicates whether the datapath should avoid masquerading connections from this IP when the cluster is in tunneling mode.
```
We're primarily talking about Pod-to-world connections here, correct? Why does that only affect overlay-routing mode - don't we have to make the same decision ("given a src and dst, should we masquerade?") in native-routing mode?

fristonio commented on pkg/endpoint/endpoint.go:
```diff
@@ -98,6 +98,11 @@ const (
+	// PropertySkipMasqueradeV4 will mark the endpoint to skip IPv4 masquerade.
+	PropertySkipMasqueradeV4 = "property-skip-masquerade-v4"
+	// PropertySkipMasqueradeV6 will mark the endpoint to skip IPv6 masquerade.
+	PropertySkipMasqueradeV6 = "property-skip-masquerade-v6"
```
Thanks, I didn't know we can specify different pools for ipv4 and ipv6, TIL.

So, its possible to have different masquerade setting for ipv4 and ipv6. Does it make sense to propagate the same semantics to the new annotation being added here? So, to have explicit annotations for  ipv4 and ipv6 masquerading instead of a single one on the pool?

qmonnet commented on Documentation/network/concepts/ipam/multi-pool.rst:
```diff
@@ -360,6 +360,25 @@ PodCIDRs allocated from ``CiliumPodIPPools`` can be announced to the network by
+node they are on.
+
+This can be achieved when eBPF-based masquerading is enabled through the flag
+``--only-masquerade-default-pool``, which will disable masquerading for all non-default
```
Nit:

```suggestion
``--only-masquerade-default-pool``, which disables masquerading for all non-default
```

qmonnet reviewed (changes requested):
Thanks! I just looked at the docs, I've got minor style suggestions (inline below).

alimehrabikoshki commented on Documentation/network/concepts/ipam/multi-pool.rst:
```diff
@@ -360,6 +360,25 @@ PodCIDRs allocated from ``CiliumPodIPPools`` can be announced to the network by
+
+Changing either the flag or the annotation after a pod has been allocated an
+IP will not change masquerade behaviour for that pod until it has re-scheduled.
+
```
>Note that given we depend on BPF masquerading, there should be some validation in the controlplane that confirms whether all the needed config options are set.

Good call, I will add an error return value to newIPAddressManager which will fail when my flag is set but BPF masq is disabled.

>This could be part of the PR description, and/or summarized here in the docs.

Sure, I'll add some more colour to the PR description

qmonnet commented on Documentation/network/concepts/ipam/multi-pool.rst:
```diff
@@ -360,6 +360,30 @@ PodCIDRs allocated from ``CiliumPodIPPools`` can be announced to the network by
+masqueraded as the IP of the node it is on.
+
+It is not always possible to identify pods that should not be masqueraded with just destination
+IPs (via --ipvX-native-routing-cidr flag or ip-masq-agent rules) as there might be overlap
```
Looks good thanks, I'd just mark these terms below as literals (I guess the spell checker may complain otherwise)

```suggestion
IPs (via ``--ipvX-native-routing-cidr`` flag or ``ip-masq-agent`` rules) as there might be overlap
```

gandro commented on pkg/ipam/cell/cell.go:
```diff
@@ -58,16 +82,38 @@ type ipamParams struct {
+func newIPAddressManager(params ipamParams, c ipamConfig) (*ipam.IPAM, error) {
+	if c.OnlyMasqueradeDefaultPool && !params.AgentConfig.EnableBPFMasquerade {
+		return nil, fmt.Errorf("--only-masquerade-default-pool requires --enable-bpf-masquerade to be enabled")
+	}
```
nit: This if statement could be moved into a separate "validateConfig" function that is invoked via `cell.Invoke` - but it's fine here as is.

alimehrabikoshki commented on pkg/ipam/multipool.go:
```diff
@@ -717,6 +739,24 @@ func (m *multiPoolManager) allocateNext(owner string, poolName Pool, family Fami
 		return nil, &ErrPoolNotReadyYet{poolName: poolName, family: family}
 	}
+	skipMasq := false
+	if poolName != PoolDefault() {
```
That was intentional, but I'm not married to the idea. I was thinking if you wanted default pool to be not masqueraded, you'd disable masquerading altogether but there *might* be the case where default is not masqueraded but other pools are? What do you think?

devodev commented:
/test

gandro commented:
/test

julianwiedmann commented:
Looks like we have CI fallout in the [multipool workflow](https://github.com/cilium/cilium/actions/workflows/conformance-multi-pool.yaml?query=branch%3Amain) from this change, although it's not clear to me why CI on the PR itself didn't catch it:

[example](https://github.com/cilium/cilium/actions/runs/20269078811/job/58199772701#step:18:636)
> time=2025-12-16T13:12:45.808654792Z level=error source=/go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/cell/lifecycle.go:129 msg="Start hook failed" function="cmd.daemonLegacyInitialization.func1 (cmd/daemon_main.go:1294) (agent.controlplane.daemon)" error="daemon configuration failed: Unable to allocate router IP for family ipv4: IP pool 'default' not found in stateDB table" (1 occurrences)

gandro commented:
Seems like it's a race. But yeah, indeed strange that we didn't catch it in this PR. We need to check in `multiPoolManager.waitForAllPools` if the StateDB table also contains the pool, before we return from `NewIPAM`
